Adding a CI for using Replay ICs with an offset hour of 3

NeilBarton-NOAA opened the pull request:
<!-- PLEASE READ -->
<!-- Any PRs not following this template will be closed -->
<!--
This PR adds a CI test that includes the use of the replay ICs that start with a +3Z offset hour. The replay ICs are used within the SFS and GEFS testing and this PR will insure these forecasts do not fail when other PRs are merged.      

-->
# Description
This PR adds a CI testing using C96mx100 resolution and the S2SWA app. 2 perturb members are included with the control member. 

Currently, this PR is dependent on 
https://github.com/NOAA-EMC/global-workflow/pull/2778
https://github.com/NOAA-EMC/global-workflow/pull/2755

as the above PRs ensure the product tasks are being triggered in the workflow. 

I could likely bypass the dependency https://github.com/NOAA-EMC/global-workflow/pull/2755 if desired. 

I will keep this PR as a draft until further discussion on the dependencies. 

Current test ICs are located on hera at
/scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/REPLAY_ICs/CI/2020110100


# Type of change
- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
This is currently being tested on HERA. 
- Forecast-only on Hera


# Checklist
- [ ] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

NeilBarton-NOAA committed:
initial creation of CI using replay ICs

NeilBarton-NOAA committed:
change name

NeilBarton-NOAA committed:
making an S2SWA case

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into gefs_replay_ci

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into gefs_replay_ci

NeilBarton-NOAA committed:
start of CI for GEFS replays # Please enter the commit message for your
changes. Lines starting

NeilBarton-NOAA committed:
Merge branch 'develop' into gefs_replay_ci

NeilBarton-NOAA committed:
removing shell check errors

NeilBarton-NOAA committed:
Merge branch 'gefs_replay_ci' of https://github.com/NeilBarton-NOAA/global-workflow into gefs_replay_ci

NeilBarton-NOAA committed:
fixes for wave post processing

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into gefs_replay_ci

NeilBarton-NOAA committed:
Merge remote-tracking branch 'origin/develop' into gefs_replay_ci

NeilBarton-NOAA committed:
add BREAKPOINTS variable

NeilBarton-NOAA commented:
ready for review. This will have a conflict with @KateFriedman-NOAA 's PR https://github.com/NOAA-EMC/global-workflow/pull/2651

NeilBarton-NOAA committed:
fix shell checks

KateFriedman-NOAA commented on parm/config/gefs/config.ufs:
```diff
@@ -328,7 +328,7 @@ if [[ "${skip_mom6}" == "false" ]]; then
       MOM6_RIVER_RUNOFF='False'
       eps_imesh="2.5e-1"
-      TOPOEDITS="ufs.topo_edits_011818.nc"
+      TOPOEDITS="topo_edits_011818.nc"
```
Should this still include the "ufs." bit, especially after issue #2812 was completed?

WalterKolczynski-NOAA replied:
I think we still need this change.

NeilBarton-NOAA replied:
code updated with ufs.topo

EricSinsky-NOAA commented on ci/cases/yamls/gefs_replay_ci.yaml:
```diff
@@ -0,0 +1,7 @@
+base:
+  REPLAY_ICS: "YES"
+  FCST_BREAKPOINTS: ""
+  BASE_CPLIC: {{ 'TOPICDIR' | getenv }}/REPLAY_ICs/CI
```
I think we'll need to have the replay IC data in the relevant directories on each platform. For example, on WCOSS2, I think we can have the replay IC data here: `/lfs/h2/emc/global/noscrub/emc.global/data/ICSDIR/`

KateFriedman-NOAA replied:
Yes. You can pass replay ICs to me and I will pull them into the `BASE_IC`/`../ICSDIR/` locations on each supported platform.

FYI, now that #2651 has been merged, you'll need to rename `BASE_CPLIC` to `BASE_IC`. See that PR for other changes that may need to be incorporated into this PR.

EricSinsky-NOAA replied:
Thank you, @KateFriedman-NOAA. The replay ICs can be found on Hera: /scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/REPLAY_ICs/CI/2020110100

@NeilBarton-NOAA Please let us know if this is the incorrect location.

KateFriedman-NOAA replied:
@EricSinsky-NOAA I have copied the replay ICs here on Hera: `/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96C96mx100`

Let me know if the `C96C96mx100` folder name is correct, I got those values from the new `ci/cases/pr/C96_S2SWA_gefs_replay_ics.yaml` that this PR is adding. Also let me know if the ICs look good underneath and we'll sync them to the other platforms.

Lastly, I notice a couple files are size 0, is that intended?
```
[role.glopara@hfe12 C96C96mx100]$ ll gefs.20201101/00/mem000/* | grep " 0 Jun"
-rw-r--r-- 1 role.glopara global          0 Jun 12 17:03 20201101.030000.ca_data.nc
-rw-r--r-- 1 role.glopara global          0 Jun 12 17:03 20201101.030000.coupler.res
```

NeilBarton-NOAA replied:
@KateFriedman-NOAA Could you use just one C96? i.e., 

/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100

KateFriedman-NOAA replied:
@NeilBarton-NOAA Yes, done: `/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100`

BTW, I notice that all of the symlinks from mem001 to mem010 that point to mem000 have double `../../mem###/TYPE` in the paths (e.g. `../../mem000/ice/../../mem000/ice`):
```
[role.glopara@hfe12 ICSDIR]$ ll /scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/REPLAY_ICs/CI/2020110100/mem001/ice/  
total 4
lrwxrwxrwx 1 Neil.Barton stmp 67 Jun 12 17:04 20201101.030000.cice_model.res.nc -> ../../mem000/ice/../../mem000/ice/20201101.030000.cice_model.res.nc
```

NeilBarton-NOAA replied:
@KateFriedman-NOAA the double soft links are due to the change in expected directory structure for the stage_ic task. I just soft linked the old structure to the new structure

EricSinsky-NOAA commented on ci/cases/yamls/gefs_replay_ci.yaml:
```diff
@@ -0,0 +1,7 @@
+defaults:
+  !INC {{ HOMEgfs }}/parm/config/gefs/yaml/defaults.yaml
+base:
+  REPLAY_ICS: "YES"
```
Some things to consider adding for this CI case:
```
FHMAX_GFS: 48
FHMAX_HF_GFS: 24
DO_EXTRACTVARS: "YES"         
FHOUT_HF_GFS: 3
```
If we are to add `FHOUT_HF_GFS` to `ci/cases/yamls/gefs_replay_ci.yaml`, a default `FHOUT_HF_GFS` will need to be added to `{{ HOMEgfs }}/parm/config/gefs/yaml/defaults.yaml`, and `FHOUT_HF_GFS=@FHOUT_HF_GFS@` will need to replace `FHOUT_HF_GFS=1` in `config.base`.

NeilBarton-NOAA replied:
Thanks, I'm currently testing.

EricSinsky-NOAA replied:
If our aim is to make this CI test like the reforecast, I also suggest setting `FHOUT_OCN_GFS` and `FHOUT_ICE_GFS` to 24 in `gefs_replay_ci.yaml`.  This would also mean adding default values for these variables in `{{ HOMEgfs }}/parm/config/gefs/yaml/defaults.yaml `and modifying `config.base` accordingly.

NeilBarton-NOAA replied:
@EricSinsky-NOAA Could you make a PR into my branch?

EricSinsky-NOAA replied:
Sure. Just created a [PR](https://github.com/NeilBarton-NOAA/global-workflow/pull/7) into your branch. I can add these changes in my PR.

NeilBarton-NOAA committed:
edit defult config.efc to be close to re-forecasting

EricSinsky-NOAA committed:
Fix bug in variable check

Fix a bug in a variable check in the extractvars task where parenthesis in the variable name will prompt a false warning message.

EricSinsky-NOAA committed:
Add more settings to GEFS replay CI test

EricSinsky-NOAA committed:
Add atmos variables requested for reforeacst

EricSinsky-NOAA committed:
Reorganize TSNOWP in paramlist

EricSinsky-NOAA committed:
Extract ice variables from history file instead

Ice variables are extracted from the history files instead of the product files. This is necessary because there are certain ice variables that are needed that are only available in the history files and not the product files.

EricSinsky-NOAA committed:
Fix typo

EricSinsky-NOAA committed:
Update TCDC variable to correct name

NeilBarton-NOAA committed:
Merge pull request #7 from EricSinsky-NOAA/gefs_replay_ci_pr

Add suggested patches to gefs_replay_ci

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into gefs_replay_ci

NeilBarton-NOAA committed:
edits to stage_ic for replay perturbations

NeilBarton-NOAA committed:
adding icsdir to pr CI yaml

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into gefs_replay_ci

EricSinsky-NOAA commented on ush/extractvars_tools.sh:
```diff
@@ -17,7 +17,7 @@
   mapfile -t requestedvar_in_allgrb2file_arr < "${requestedvar_in_allgrb2file}"
   while read -r vari; do
-    if [[ ! ${requestedvar_in_allgrb2file_arr[*]} =~ ${vari} ]] ;then
+    if [[ ! ${requestedvar_in_allgrb2file_arr[*]} =~ "${vari}" ]] ;then
```
@WalterKolczynski-NOAA, can this be shellcheck disabled? The only way to make shellnorms happy here is to remove quotes around  `${vari}`. If these quotes are removed, we would have to put a backslash in front of every regex special character (namely parenthesis) in every parmlist file. Since we want every regex special character to be treated as a literal in these parm files, I think it might be ok to put quotes here.

WalterKolczynski-NOAA replied:
As long as the substitution still happens, yes. I was worried it would treat `${vari}` itself as literal.

WalterKolczynski-NOAA replied:
To suppress the warning
```suggestion
    # shellcheck disable=SC2076
    if [[ ! ${requestedvar_in_allgrb2file_arr[*]} =~ "${vari}" ]] ;then
```

EricSinsky-NOAA replied:
Thanks @WalterKolczynski-NOAA. Accidentally put the shellcheck disable in the wrong place. I sent a PR to @NeilBarton-NOAA's `gefs_replay_ci` branch that fixes this.

EricSinsky-NOAA committed:
Fix ensstat failures

NeilBarton-NOAA committed:
updates for products

EricSinsky-NOAA committed:
Add shellcheck disable and skip f000 for ocn/ice

Shellcheck disable has been added. Also, f000 is skipped in extractvars for ocn and ice since there are no products for f000 for ocn and ice. This change will remove warning messages.

EricSinsky-NOAA committed:
Revert "Fix ensstat failures"

This reverts commit 4985361113b4f7e3134a791ff8fdb5a7969644d7.

EricSinsky-NOAA committed:
Merge remote-tracking branch 'nb/gefs_replay_ci' into gefs_replay_ci_pr

NeilBarton-NOAA committed:
Merge pull request #8 from EricSinsky-NOAA/gefs_replay_ci_pr

Fix ensstat failures

EricSinsky-NOAA committed:
Merge remote-tracking branch 'nb/gefs_replay_ci' into gefs_replay_ci_pr

EricSinsky-NOAA committed:
Move shellcheck disable

EricSinsky-NOAA committed:
Remove shellcheck disable

NeilBarton-NOAA commented:
in my opinion, this is ready for further review @WalterKolczynski-NOAA @aerorahul

WalterKolczynski-NOAA commented:
> in my opinion, this is ready for further review @WalterKolczynski-NOAA @aerorahul

I thought we were redirecting this to the new `feature/gefs_reforecast` branch. You should be able to just change the target branch for this PR (or I can do it).

EricSinsky-NOAA commented:
> I thought we were redirecting this to the new `feature/gefs_reforecast` branch. You should be able to just change the target branch for this PR (or I can do it).

@WalterKolczynski-NOAA Will we be able to run CI tests with `feature/gefs_reforecast` as the target branch? If yes, I am fine with  `feature/gefs_reforecast` being the target of this PR.

WalterKolczynski-NOAA commented:
> > I thought we were redirecting this to the new `feature/gefs_reforecast` branch. You should be able to just change the target branch for this PR (or I can do it).
> 
> @WalterKolczynski-NOAA Will we be able to run CI tests with `feature/gefs_reforecast` as the target branch? If yes, I am fine with `feature/gefs_reforecast` being the target of this PR.

Should be able to.

WalterKolczynski-NOAA commented:
Changed the target to `feature/gefs_reforecast`

WalterKolczynski-NOAA commented on scripts/exgfs_wave_post_pnt.sh:
```diff
@@ -252,7 +252,9 @@ source "${USHgfs}/preamble.sh"
     ${NLN} mod_def.$waveuoutpGRD mod_def.ww3
-    HMS="${cyc}0000"
+    #export OFFSET_START_HOUR=$( printf "%02d" ${half_assim} )
+    sec=$( printf "%02d" $(( cyc + OFFSET_START_HOUR )) )
+    HMS="${sec}0000"
```
Using `sec` for an hour is extremely confusing
```suggestion
    hh=$( printf "%02d" $(( cyc + OFFSET_START_HOUR )) )
    HMS="${hh}0000"
```

WalterKolczynski-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -324,10 +324,12 @@ def atmos_ensstat(self):
     def wavepostsbs(self):
         deps = []
+        dtg_prefix = "@Y@m@d.@H0000"
+        offset = str(self._configs['base']['OFFSET_START_HOUR']).zfill(2) + ":00:00"
```
Use `aerosol_init` from `gfs_tasks.py` as a model.
```suggestion
        interval = self._configs['base']['OFFSET_START_HOUR']
        offset = timedelta_to_HMS(interval)
```

NeilBarton-NOAA commented:
@aerorahul and I talked about this going into dev. Otherwise updates going into dev have a great potential of breaking GEFS and SFS runs

aerorahul commented on ush/ocnice_extractvars.sh:
```diff
@@ -21,15 +21,23 @@ comout_rfcst_prod_ocnice=${6}
   if [[ ${component_name} == "ocn" ]]; then
-    infile=${COMIN_OCEAN_NETCDF}/${datares}/${RUN}.ocean.t${cyc}z.${datares}.f${fnh}.nc
+    if [[ "${datares}" == "native" ]]; then
+      infile=${COMIN_OCEAN_HISTORY}/${RUN}.ocean.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc
```
```suggestion
      infile="${COMIN_OCEAN_HISTORY}/${RUN}.ocean.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc"
```

WalterKolczynski-NOAA commented:
Changed the target back to develop.

aerorahul commented on ush/ocnice_extractvars.sh:
```diff
@@ -21,15 +21,23 @@ comout_rfcst_prod_ocnice=${6}
+    if [[ "${datares}" == "native" ]]; then
+      infile=${COMIN_OCEAN_HISTORY}/${RUN}.ocean.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc
+    else
+      infile=${COMIN_OCEAN_NETCDF}/${datares}/${RUN}.ocean.t${cyc}z.${datares}.f${fnh}.nc
```
```suggestion
      infile="${COMIN_OCEAN_NETCDF}/${datares}/${RUN}.ocean.t${cyc}z.${datares}.f${fnh}.nc"
```
What is the value of `datares` supposed to be here? This is the interpolated grid correct? Should this be `grid_res`; `1p00`, 0p25` etc?

EricSinsky-NOAA replied:
For the reforecast, we aim to have `datares` be `1p00` for ocean and `native` (native resolution) for ice. This means we are extracting ocean products with an interpolated grid from `COMIN_OCEAN_NETCDF`. We are extracting ice products with a native grid from `COMIN_ICE_HISTORY`. We are extracting ice variables from `COMIN_ICE_HISTORY` because all the variables that we need to extract are available there.

aerorahul commented on ush/ocnice_extractvars.sh:
```diff
@@ -21,15 +21,23 @@ comout_rfcst_prod_ocnice=${6}
   elif [[ ${component_name} == "ice" ]]; then
-    infile=${COMIN_ICE_NETCDF}/${datares}/${RUN}.ice.t${cyc}z.${datares}.f${fnh}.nc
+    if [[ "${datares}" == "native" ]]; then
+      infile=${COMIN_ICE_HISTORY}/${RUN}.ice.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc
```
```suggestion
      infile="${COMIN_ICE_HISTORY}/${RUN}.ice.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc"
```

aerorahul commented on ush/ocnice_extractvars.sh:
```diff
@@ -21,15 +21,23 @@ comout_rfcst_prod_ocnice=${6}
+    if [[ "${datares}" == "native" ]]; then
+      infile=${COMIN_ICE_HISTORY}/${RUN}.ice.t${cyc}z.${fhout_ocnice}hr_avg.f${fnh}.nc
+    else
+      infile=${COMIN_ICE_NETCDF}/${datares}/${RUN}.ice.t${cyc}z.${datares}.f${fnh}.nc
```
```suggestion
      infile="${COMIN_ICE_NETCDF}/${datares}/${RUN}.ice.t${cyc}z.${datares}.f${fnh}.nc"
```

aerorahul reviewed (commented):
some minor comments.

EricSinsky-NOAA committed:
Add double quotes in ocnice_extractvars

Double quotes were added to address a reviewer's suggestion.

EricSinsky-NOAA commented:
@aerorahul I have addressed your review comments related to the extractvars task. With Neil's permission, I have pushed these changes directly to his branch.

NeilBarton-NOAA committed:
Merge branch 'develop' into gefs_replay_ci

NeilBarton-NOAA committed:
Merge branch 'NOAA-EMC:develop' into gefs_replay_ci

EricSinsky-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -324,10 +324,12 @@ def atmos_ensstat(self):
         deps.append(rocoto.add_dependency(dep_dict))
         dependencies = rocoto.create_dependency(dep=deps)
 
+
```
Removing this space should resolve the pynorms error. 

```suggestion
```

NeilBarton-NOAA committed:
updates for new MOM6 100 input file

NeilBarton-NOAA committed:
Update workflow/rocoto/gefs_tasks.py

Co-authored-by: Eric Sinsky - NOAA <48259628+EricSinsky-NOAA@users.noreply.github.com>

NeilBarton-NOAA committed:
fix archiving until archiving PR is ready

NeilBarton-NOAA committed:
Merge branch 'gefs_replay_ci' of https://github.com/NeilBarton-NOAA/global-workflow into gefs_replay_ci

NeilBarton-NOAA committed:
Merge branch 'develop' into gefs_replay_ci

NeilBarton-NOAA committed:
change back to ufs.top edits

NeilBarton-NOAA committed:
Merge branch 'gefs_replay_ci' of https://github.com/NeilBarton-NOAA/global-workflow into gefs_replay_ci

NeilBarton-NOAA committed:
Update scripts/exgfs_wave_post_pnt.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

NeilBarton-NOAA commented:
@WalterKolczynski-NOAA @aerorahul This PR is ready for further review/CI testing.

WalterKolczynski-NOAA commented on ush/parsing_namelists_MOM6.sh:
```diff
@@ -41,6 +41,10 @@ cat input.nml
 # Prepare local variables for use in MOM_input.IN from UFSWM
 # The ones already defined are left commented as a reminder
+# == MOM options to start from coarsed grained restarts, set to off by default
+local MOM6_INIT_FROM_Z=${MOM6_INIT_FROM_Z:-True}
```
This should be False to match the current default
```suggestion
local MOM6_INIT_FROM_Z=${MOM6_INIT_FROM_Z:-False}
```

NeilBarton-NOAA replied:
The default is True, see line 1016

https://github.com/NeilBarton-NOAA/ufs-weather-model/blob/develop/tests/default_vars.sh

It is set to False for use with the coarse grained/interpolated MOM6 restarts.

WalterKolczynski-NOAA replied:
Okay, they have the wrong comment in the template then: https://github.com/ufs-community/ufs-weather-model/blob/develop/tests/parm/MOM_input_050.IN#L309

WalterKolczynski-NOAA commented on ush/parsing_namelists_MOM6.sh:
```diff
@@ -41,6 +41,10 @@ cat input.nml
 # Prepare local variables for use in MOM_input.IN from UFSWM
 # The ones already defined are left commented as a reminder
+# == MOM options to start from coarsed grained restarts, set to off by default
+local MOM6_INIT_FROM_Z=${MOM6_INIT_FROM_Z:-True}
+local MOM6_WARMSTART_FILE=${MOM6_WARMSTART_FILE:-"none"}
+local MOM6_INIT_UV=${MOM6_INIT_UV:-"zero"}
```
NB: These variables are not in the MOM namelists for 0p25 or 5p00; only 0p50 and 1p00.

NeilBarton-NOAA replied:
correct. They are used with taking 0.25 MOM restarts and interpolating them to a lower resolution. Do you recommend them being at another location?

WalterKolczynski-NOAA replied:
Maybe just add a comment nothing they are only used for 0p50 and 1p00?

NeilBarton-NOAA replied:
completed

WalterKolczynski-NOAA commented on parm/archive/gefs_arcdir.yaml.j2:
```diff
@@ -21,7 +21,7 @@
     {% set ensstat_files = [] %}
     {% if path_exists(COMIN_ATMOS_ENSSTAT_1p00) %}
-        {% for fhr in range(FHMIN_GFS, FHMAX_GFS + FHOUT_GFS, FHOUT_GFS) %}
+        {% for fhr in range(6, FHMAX_GFS + FHOUT_GFS, FHOUT_GFS) %}
```
Should this really be hard-coded?

NeilBarton-NOAA replied:
@EricSinsky-NOAA do you have a recommendation? This will be changed with the archiving PR.

EricSinsky-NOAA replied:
@NeilBarton-NOAA Please see my review comments below. It turns out `parm/archive/gefs_arcdir.yaml.j2` is not being modified in Anton's current PR, so we might need to make that correction here.

EricSinsky-NOAA commented on parm/archive/gefs_arcdir.yaml.j2:
```diff
@@ -21,7 +21,7 @@
 {% if RUN == "gefs" %}
```
To add the lines below, `REPLAY_ICS` and `OFFSET_START_HOUR` will need to be added to `keys` in `scripts/exglobal_archive.py`, but this is already being done in Anton's PR. 

```suggestion
{% if REPLAY_ICS == "YES" %}
    {% set ofst_hr = FHOUT_GFS %}
{% else %}
    {% set ofst_hr = FHMIN_GFS %}
{% endif %}

{% if RUN == "gefs" %}
```

EricSinsky-NOAA commented on parm/archive/gefs_arcdir.yaml.j2:
```diff
@@ -21,7 +21,7 @@
     {% set ensstat_files = [] %}
     {% if path_exists(COMIN_ATMOS_ENSSTAT_1p00) %}
-        {% for fhr in range(FHMIN_GFS, FHMAX_GFS + FHOUT_GFS, FHOUT_GFS) %}
+        {% for fhr in range(6, FHMAX_GFS + FHOUT_GFS, FHOUT_GFS) %}
```
```suggestion
        {% for fhr in range(ofst_hr, FHMAX_GFS + FHOUT_GFS, FHOUT_GFS) %}
```

NeilBarton-NOAA committed:
Update parm/archive/gefs_arcdir.yaml.j2

Co-authored-by: Eric Sinsky - NOAA <48259628+EricSinsky-NOAA@users.noreply.github.com>

NeilBarton-NOAA committed:
add comments for coarse grained restarts

NeilBarton-NOAA committed:
Merge branch 'gefs_replay_ci' of https://github.com/NeilBarton-NOAA/global-workflow into gefs_replay_ci

EricSinsky-NOAA committed:
Make archive job work with replay run

EricSinsky-NOAA committed:
Add Foundation SST to group-a products

Foundation SST has been moved to the group-a products so that it can be included in ensstat.

EricSinsky-NOAA committed:
Modify gefs_arcdir yaml

The gefs_arcdir yaml has been modified so that it can run with both replay and non-replay runs.

EricSinsky-NOAA committed:
Fix gefs_arcdir yaml

EricSinsky-NOAA reviewed (dismissed):
All looks good to me. I ran this on WCOSS2 and all tasks were successful, including the archive task.

NeilBarton-NOAA committed:
updates for stage_ic

NeilBarton-NOAA committed:
Merge branch 'develop' into gefs_replay_ci

NeilBarton-NOAA commented:
@KateFriedman-NOAA I update the stage ic yamls with the default directory structure, You'll have to pick up the ICs again at 

hera:/scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/REPLAY_ICs/CI

Everything else is working on my side

KateFriedman-NOAA commented:
> @KateFriedman-NOAA I update the stage ic yamls with the default directory structure, You'll have to pick up the ICs again at
> 
> hera:/scratch2/NCEPDEV/stmp3/Neil.Barton/ICs/REPLAY_ICs/CI

@NeilBarton-NOAA I've pulled the updated IC set into `ICSDIR` on Hera. See here: `/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/20240610`

I pulled the files through the symlinks I saw within your set (`rsync -azvL`), so as to not have any symlinks to personal files. I also placed it under a timestamp folder. Let me know if the ICs look good on Hera and we'll sync them to the other platforms. After we sync them I will also remove the prior set that I pulled in that wasn't yet under the timestamp folder (`/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/gefs.20201101`).

NeilBarton-NOAA committed:
Merge branch 'develop' into gefs_replay_ci

EricSinsky-NOAA committed:
Merge remote-tracking branch 'EMC/develop' into gefs_replay_ci

EricSinsky-NOAA commented:
Conflicts have been resolved. I think this PR is ready to move forward.

aerorahul commented:
> Checkout **Failed** on Hera in Build# 2: Could not perform submodule update

@TerrenceMcGuinness-NOAA @KateFriedman-NOAA 
What are we dealing with here?  Seems like Hera issues.  Do we need to open a ticket w/ Hera admins?

WalterKolczynski-NOAA commented:
> > Checkout **Failed** on Hera in Build# 2: Could not perform submodule update
> 
> @TerrenceMcGuinness-NOAA @KateFriedman-NOAA What are we dealing with here? Seems like Hera issues. Do we need to open a ticket w/ Hera admins?

There have been intermittent but persistent issues checking out gdas on Hera for at least a week now (specifically git-lfs).

@WalterKolczynski-NOAA We have the same issue on Gaea.  The `git-lfs` is timing out in the "Nodes" and we an not reproducing the error directly at the prompts outside of the Jenkins Nodes.  I am looking into increasing the time limits set by the Jenkins **scm** plugin. ~T.McG

WalterKolczynski-NOAA commented:
```
Submodule 'pkg/CVMix-src' (https://github.com/mom-ocean/CVMix-src.git) registered for path 'sorc/gdas.cd/sorc/soca/mom6/MOM6/pkg/CVMix-src'
Submodule 'pkg/GSW-Fortran' (https://github.com/mom-ocean/GSW-Fortran.git) registered for path 'sorc/gdas.cd/sorc/soca/mom6/MOM6/pkg/GSW-Fortran'
Cloning into '/scratch1/NCEPDEV/global/CI/2788/gfs/sorc/gdas.cd/sorc/soca/mom6/MOM6/pkg/CVMix-src'...
Cloning into '/scratch1/NCEPDEV/global/CI/2788/gfs/sorc/gdas.cd/sorc/soca/mom6/MOM6/pkg/GSW-Fortran'...
fatal: the remote end hung up unexpectedly
error: git-lfs filter-process died of signal 15
fatal: Unable to checkout 'cd66505007b1559d79cb158bd6dc018a3943c1e7' in submodule path 'sorc/gdas.cd/sorc/ufo'
fatal: Failed to recurse into submodule path 'sorc/gdas.cd'

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2846)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:2185)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$7.lambda$execute$0(CliGitAPIImpl.java:1573)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at com.google.common.util.concurrent.DirectExecutorService.execute(DirectExecutorService.java:51)
	at java.base/java.util.concurrent.ExecutorCompletionService.submit(ExecutorCompletionService.java:184)
	at org.jenkinsci.plugins.gitclient.cgit.GitCommandsExecutor.submitRemainingCommand(GitCommandsExecutor.java:77)
	at org.jenkinsci.plugins.gitclient.cgit.GitCommandsExecutor.invokeAll(GitCommandsExecutor.java:70)
Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to Hera-EMC
		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1826)
		at hudson.remoting.UserRequest$ExceptionResponse.retrieve(UserRequest.java:356)
		at hudson.remoting.Channel.call(Channel.java:1042)
		at PluginClassLoader for git-client//org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.execute(RemoteGitImpl.java:153)
		at jdk.internal.reflect.GeneratedMethodAccessor421.invoke(Unknown Source)
		at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.base/java.lang.reflect.Method.invoke(Method.java:569)
		at PluginClassLoader for git-client//org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler.invoke(RemoteGitImpl.java:138)
		at PluginClassLoader for git-client/jdk.proxy26/jdk.proxy26.$Proxy109.execute(Unknown Source)
		at PluginClassLoader for git//hudson.plugins.git.extensions.impl.SubmoduleOption.onCheckoutCompleted(SubmoduleOption.java:196)
		at PluginClassLoader for git//hudson.plugins.git.GitSCM._checkout(GitSCM.java:1395)
		at PluginClassLoader for git//hudson.plugins.git.GitSCM.checkout(GitSCM.java:1277)
		at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:136)
		at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:101)
		at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:88)
		at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
		at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
		at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
		at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
		at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
		at java.base/java.lang.Thread.run(Thread.java:840)
Caused: hudson.plugins.git.GitException
	at org.jenkinsci.plugins.gitclient.cgit.GitCommandsExecutor.checkResult(GitCommandsExecutor.java:89)
	at org.jenkinsci.plugins.gitclient.cgit.GitCommandsExecutor.invokeAll(GitCommandsExecutor.java:69)
	at org.jenkinsci.plugins.gitclient.cgit.GitCommandsExecutor.invokeAll(GitCommandsExecutor.java:47)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$7.execute(CliGitAPIImpl.java:1576)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:170)
	at org.jenkinsci.plugins.gitclient.RemoteGitImpl$CommandInvocationHandler$GitCommandMasterToSlaveCallable.call(RemoteGitImpl.java:161)
	at hudson.remoting.UserRequest.perform(UserRequest.java:211)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:377)
	at hudson.remoting.InterceptingExecutorService.lambda$wrap$0(InterceptingExecutorService.java:78)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at hudson.remoting.Engine$1.lambda$newThread$0(Engine.java:137)
	at java.base/java.lang.Thread.run(Thread.java:842)
Caused: java.io.IOException: Could not perform submodule update
	at PluginClassLoader for git//hudson.plugins.git.extensions.impl.SubmoduleOption.onCheckoutCompleted(SubmoduleOption.java:201)
	at PluginClassLoader for git//hudson.plugins.git.GitSCM._checkout(GitSCM.java:1395)
	at PluginClassLoader for git//hudson.plugins.git.GitSCM.checkout(GitSCM.java:1277)
	at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:136)
	at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:101)
	at PluginClassLoader for workflow-scm-step//org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:88)
	at PluginClassLoader for workflow-step-api//org.jenkinsci.plugins.workflow.steps.SynchronousNonBlockingStepExecution.lambda$start$0(SynchronousNonBlockingStepExecution.java:47)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
ERROR: Maximum checkout retry attempts reached, aborting
ERROR: Timeout after 10 minutes
```

WalterKolczynski-NOAA commented:
ICSDIR not being populated correctly on WCOSS2 (possibly elsewhere too?):
```
+++ config.stage_ic[10]: export ICSDIR=UNDEFINED/REPLAY_ICs/CI
```

EricSinsky-NOAA commented on ci/cases/pr/C96_S2SWA_gefs_replay_ics.yaml:
```diff
@@ -0,0 +1,19 @@
+  idate: 2020110100
+  edate: 2020110100
+  yaml: {{ HOMEgfs }}/ci/cases/yamls/gefs_replay_ci.yaml
+  icsdir: {{ 'TOPICDIR' | getenv }}/REPLAY_ICs/CI
```
This should be `ICSDIR_ROOT`. I do not believe `TOPICDIR` is defined anywhere in global-workflow, which is why the CI test failed on WCOSS2. We'll need to verify that `REPLAY_ICs/CI` is in the following directories on each platform.
Hera:` /scratch1/NCEPDEV/global/glopara/data/ICSDIR`
WCOSS2:` /lfs/h2/emc/global/noscrub/emc.global/data/ICSDIR`
Orion: `/work/noaa/global/glopara/data/ICSDIR`
Hercules: `/work/noaa/global/glopara/data/ICSDIR`
 

```suggestion
  icsdir: {{ 'ICSDIR_ROOT' | getenv }}/REPLAY_ICs/CI
```

NeilBarton-NOAA committed:
point to CI ICs

NeilBarton-NOAA committed:
Merge branch 'gefs_replay_ci' of https://github.com/NeilBarton-NOAA/global-workflow into gefs_replay_ci

NeilBarton-NOAA commented:
I edited the YAML to point to the CI ICs instead of my personal ICs.  It's now working on hera

EricSinsky-NOAA commented on ci/cases/pr/C96_S2SWA_gefs_replay_ics.yaml:
```diff
@@ -0,0 +1,19 @@
+  idate: 2020110100
+  edate: 2020110100
+  yaml: {{ HOMEgfs }}/ci/cases/yamls/gefs_replay_ci.yaml
+  icsdir: {{ 'ICSDIR_ROOT' | getenv }}/C96mx100/20240610
```
@NeilBarton-NOAA Should this be `gefs.20201101` instead of `20240610`? 
```suggestion
  icsdir: {{ 'ICSDIR_ROOT' | getenv }}/C96mx100/gefs.20201101
```

NeilBarton-NOAA replied:
@EricSinsky-NOAA 
on hera
ls /scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/20240610
gefs.20201031  
gefs.20201101

i.e., the ICs are correct. I don't know about the name of the 20240610 directory.

EricSinsky-NOAA replied:
@NeilBarton-NOAA Please disregard my suggested change. I realized `20240610` is a timestamp folder and within this folder is the `20201101` case. It seems this directory exists on Hera, but not the other platforms. @KateFriedman-NOAA When you get a chance, can you sync the IC data under the timestamp folder to the other platforms?

KateFriedman-NOAA replied:
Synced to WCOSS2. @WalterKolczynski-NOAA please sync to MSU.

EricSinsky-NOAA replied:
Thank you @KateFriedman-NOAA. All looks good in `icsdir` on WCOSS2.

WalterKolczynski-NOAA replied:
MSU is synchronized.

TerrenceMcGuinness-NOAA commented:
I increased the timeout settings for **scm** (Software Control Module) in the Jenkins project.   If you are ready you can restart the CI tests and see if that helps with the time outs from the submodule checkouts.

TerrenceMcGuinness-NOAA commented:
The check phase of build was unable to update some submodules.  Let me clear out all the Jenkins cache files for this work space so scm can start completely over and do new clone.

TerrenceMcGuinness-NOAA commented:
`Cannot contact Hera-EMC: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@1e281834:Hera-EMC": Remote call on Hera-EMC failed. The channel is closing down or has closed down`

Hera lost connection with the Jenkins Controller.  The connection was restored with health checking scripts and the PR CI is being relaunched.

@WalterKolczynski-NOAA NOTE that it was during the actual build stage after cloning  that this "disconnected" error occurred on Hera so it is unrelated to the prior time out issues that we believe have been resolved.

WalterKolczynski-NOAA commented:
Looks like a 'real' error this time:
```
[38;21m2024-10-02 20:37:31,785 - INFO     - oceanice_products: Copy ocean data to run directory[0m
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2788/gefs/ush/python/wxflow/fsutils.py", line 85, in cp
    shutil.copy2(source, target)
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/python-3.11.6-b6ydksr/lib/python3.11/shutil.py", line 436, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/gsi-addon-dev-rocky8/install/intel/2021.5.0/python-3.11.6-b6ydksr/lib/python3.11/shutil.py", line 256, in copyfile
    with open(src, 'rb') as fsrc:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/scratch1/NCEPDEV/global/CI/2788/RUNTESTS/COMROOT/C48_S2SWA_gefs_f918cc62/gefs.20210323/12/mem000/model/ocean/history/gefs.ocean.t12z.6hr_avg.f006.nc'
```

EricSinsky-NOAA commented on parm/config/gefs/config.base:
```diff
@@ -290,19 +299,10 @@ export NMEM_ENS=@NMEM_ENS@
 # Check if cycle is cold starting
 if [[ "${EXP_WARM_START}" = ".false." ]]; then
-  export IAU_FHROT=${OFFSET_START_HOUR}
+    export OFFSET_START_HOUR=$(( assim_freq / 2 ))
```
It looks like all CI tests passed including `C96_S2SWA_gefs_replay_ics`, except `C48_S2SWA_gefs` failed. After reading the logs, it looks like `OFFSET_START_HOUR` is set to `3` here for `C48_S2SWA_gefs`. Shouldn't `OFFSET_START_HOUR` be `0` for non-replay GEFS?

NeilBarton-NOAA replied:
I had to change this for the SFS PR, I'll update and test.

EricSinsky-NOAA replied:
Thanks @NeilBarton-NOAA. After a quick test on WCOSS2, reverting the change on line 305 resolved the issue that occurred in the `C48_S2SWA_gefs` CI test. `OFFSET_START_HOUR` is already set earlier in `config.base`.

NeilBarton-NOAA committed:
fix C48_S2SWA_gefs CI error and added HPC_ACCOUNT

NeilBarton-NOAA commented:
CI test fix updated. Please put through CI tests again

KateFriedman-NOAA commented:
> @NeilBarton-NOAA I've pulled the updated IC set into `ICSDIR` on Hera. See here: `/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/20240610`
> 
> I pulled the files through the symlinks I saw within your set (`rsync -azvL`), so as to not have any symlinks to personal files. I also placed it under a timestamp folder. Let me know if the ICs look good on Hera and we'll sync them to the other platforms. After we sync them I will also remove the prior set that I pulled in that wasn't yet under the timestamp folder (`/scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/gefs.20201101`).

@NeilBarton-NOAA Checking back on this. Please confirm the above ICs are good-to-go. Thanks!

NeilBarton-NOAA commented:
@KateFriedman-NOAA The CI tests passed and the code has been merged. The ICs are good-to-go

KateFriedman-NOAA commented:
Thanks for confirming @NeilBarton-NOAA , particularly for our records. I will proceed with deleting the non-timestamped copy (`scratch1/NCEPDEV/global/glopara/data/ICSDIR/C96mx100/gefs.20201101`).

NeilBarton-NOAA commented:
@KateFriedman-NOAA I miss understood your question. The CI points to the timestamped ICs

https://github.com/NOAA-EMC/global-workflow/blob/develop/ci/cases/pr/C96_S2SWA_gefs_replay_ics.yaml

If removing the time-stamp folder, the yaml file will need to be edited.

KateFriedman-NOAA commented:
@NeilBarton-NOAA All good, I removed the _non_-timestamped one. The timestamped one will remain and be used in the CI test.
